rust/common-graph/src/storage.rs

```diff
@@ -0,0 +1,209 @@
+use super::{CommonGraphError, PortType, Result};
+use crate::utils::is_full;
+use common_macros::NewType;
+
+#[cfg(doc)]
+use crate::Graph;
+
+/// A set of input or output ports in [`GraphData<V>`].
+pub type GraphDataInner<'a, V> = Vec<(&'a str, Option<V>)>;
+
+/// Contains a tuple of input and output ports,
+/// each containing a tuple of port name and its
+/// current value, which may or may not be set.
+#[derive(NewType, Clone)]
+#[new_type(only(Inner, Into, From))]
+pub struct GraphData<'a, V>(Vec<(GraphDataInner<'a, V>, GraphDataInner<'a, V>)>);
+
+impl<'a, V> GraphData<'a, V> {
+    /// Convert this [`GraphData<V>`] into its owned
+    /// counterpart, [`OwnedGraphData<V>`].
+    pub fn into_owned(self) -> OwnedGraphData<V> {
+        self.0
+            .into_iter()
+            .map(|(ins, outs)| {
+                (
+                    ins.into_iter().map(|(k, v)| (String::from(k), v)).collect(),
+                    outs.into_iter()
+                        .map(|(k, v)| (String::from(k), v))
+                        .collect(),
+                )
+            })
+            .collect::<Vec<_>>()
+            .into()
+    }
+}
+
+/// A set of input or output ports in [`OwnedGraphData<V>`].
+pub type OwnedGraphDataInner<V> = Vec<(String, Option<V>)>;
+
+/// An "owned" version of [`GraphData`] where the keys
+/// are [`String`]s rather than references to a [`Graph`].
+#[derive(NewType, Clone)]
+#[new_type(only(Inner, Into, From))]
+pub struct OwnedGraphData<V>(Vec<(OwnedGraphDataInner<V>, OwnedGraphDataInner<V>)>);
+
+type PortStoreMut<'store, 'a, V> = (
+    &'store Vec<(&'a str, Option<V>)>,
+    Vec<(&'a str, &'store mut Option<V>)>,
+);
+
+/// Storage for port data while traversing a [`Graph`].
+/// Used to collect data during traversal, and generate
+/// completed graph traversal representations.
+///
+/// Must be initialized up front with all port names
+/// of nodes, as node state cares about whether
+/// their inputs are "full" or not.
+#[derive(NewType)]
+#[new_type(only(Into))]
+pub struct GraphStorage<'a, V>(GraphData<'a, V>);
+
+impl<'a, V> GraphStorage<'a, V> {
+    /// Initializes storage for all nodes with the provided keys.
+    pub(crate) fn from_iter<I1, I2, I3>(iter: I1) -> Self
+    where
+        I1: IntoIterator<Item = (I2, I3)>,
+        I2: IntoIterator<Item = &'a str>,
+        I3: IntoIterator<Item = &'a str>,
+    {
+        let mut store = vec![];
+        for (inputs, outputs) in iter {
+            let inputs: Vec<_> = inputs.into_iter().map(|key| (key, None)).collect();
+            let outputs: Vec<_> = outputs.into_iter().map(|key| (key, None)).collect();
+            store.push((inputs, outputs));
+        }
+        GraphStorage(store.into())
+    }
+
+    /// Sets the value of node `index`'s `port_type` port with name
+    /// `port_name` to `value`.
+    pub(crate) fn set(
```

Comment:
What are your thoughts on memory-ceiling issues? Sufficiently large graphs will call for a lot of overhead.

Reply:
* A memory-conscious processor function can `.take()` the input values to release heap allocation.
* All the key names are already references from `Graph`.

There could be some optimizations related to index based ports and a single Vec for indexing all nodes' port data. Current approach is a compromise between overhead and remaining flexible to change logic and APIs. Maybe there are other clever approaches here, though currently negligible compared to e.g. wasm runtime memory usage